Add image dimension and file size information

kaf-lamed-beyt opened the pull request:
Closes https://github.com/zed-industries/zed/issues/21281

@jansol, kindly take a look when you're free.

![image](https://github.com/user-attachments/assets/da9a54fa-6284-4012-a243-7e355a5290d3) 

Release Notes:

- Added dimensions and file size information for images.

kaf-lamed-beyt committed:
chore: show image info

kaf-lamed-beyt committed:
Merge branch 'main' of github.com:kaf-lamed-beyt/zed

kaf-lamed-beyt commented:
@cla-bot check

kaf-lamed-beyt commented:
Alright! Thank you for the feedback @Angelk90. I'll take the screenshot and share it here. 

Quick question though... which crate would I need to update to show this image data in the location you asked me to move it? I'm thinking this one: `crates/zed/src/zed/app_menus.rs`?

kaf-lamed-beyt commented:
Here's what i have

![image](https://github.com/user-attachments/assets/a79d134c-fc73-4d11-9d7d-5cca75ef3a65)

kaf-lamed-beyt commented:
Oh! Now I understand the reason for the disparity in file size. So what do you recommend we use in this case @Angelk90? the binary or decimal system? 

We can decide to accommodate both of them though. If that's okay

![image](https://github.com/user-attachments/assets/c7e9ceeb-d7f7-4714-82bb-06493f4ce465)


> @kaf-lamed-beyt : If you can show this information for the images too that would be great.

cool I'll include this too.

kaf-lamed-beyt commented:
@danilo-leal, hi 👋🏼

any idea how I can include the image metadata in the status bar?

jansol commented:
That was fast! Nice work so far!

> Anyway, an important point in my opinion is the convention of whether to use a dot or a comma to define decimals.

The decimal separator is defined by the user's locale. We should not try to be clever about it.

> any idea how I can include the image metadata in the status bar?

Don't. That won't work well if someone opens multiple images side by side. Also there is very limited space in the status bar and I would very much like to see a lot more metadata (file format; color format - channels, bit depth, subsampling, etc; comment; whether the file has color profile; ...)

As far as I've seen, most applications tend to put metadata in an overlay or a side bar. I think @iamnbutler is the one that should be consulted on that, as he is the main designer in the Zed team AIUI. You might want to arrange a collab call with him to decide on the UI design portion of this.

kaf-lamed-beyt commented:
> That was fast! Nice work so far!

Thank you, @jansol!

> Don't. That won't work well if someone opens multiple images side by side. Also there is very limited space in the status bar and I would very much like to see a lot more metadata (file format; color format - channels, bit depth, subsampling, etc; physical size/DPI; comment; whether the file has color profile; ...)

Oh! That's good then. This is a valid reason.

I'd check if the image crate can provide all these metadata you mentioned. 

What's the best way to reach @iamnbutler? Email? Twitter? Discord (I may need the discord handle)

mikayla-maki commented on crates/image_viewer/Cargo.toml:
```diff
@@ -23,3 +23,4 @@ theme.workspace = true
 ui.workspace = true
 util.workspace = true
 workspace.workspace = true
+image = "0.25.5"
```
Let's elevate this dependency, and the one in GPUI, to a workspace dependency

mikayla-maki commented on crates/image_viewer/src/image_viewer.rs:
```diff
@@ -145,6 +160,35 @@ impl Item for ImageView {
+        return Err(format!("File does not exist at path: {:?}", path));
+    }
+
+    let img = image::open(&path).map_err(|e| format!("Failed to open image: {}", e))?;
```
This and the `metadata()` call below should not be here. We don't want to block the UI thread with filesystem access. This whole function should move to `ImageItem::try_open` and `ImageItem::reload`

mikayla-maki reviewed (changes requested):
Personally, I prefer the VSCode style with the size and dimensions in the status bar. You can define a new status bar item in the `image_viewer` crate, and then add it after the cursor position indicator here: https://github.com/zed-industries/zed/blob/d7ea3ddc7cc8c26a5682c7eeb1806726f109bd5e/crates/zed/src/zed.rs#L214-L222

kaf-lamed-beyt committed:
implement image_info StatusItemView

kaf-lamed-beyt committed:
minor

kaf-lamed-beyt commented:
@mikayla-maki am i on the right path? 

I've managed to get a UI element to render in the status bar. But, I do not think the `load_metadata` fn here works as it should. I even tried logging to the console so I could see if the data was retrieved or not. 

https://github.com/zed-industries/zed/blob/b27afc46f8b0c8e1ac3a27354a3b041750565fe4/crates/project/src/image_store.rs#L135-L157

Perhaps, you could point me in the right direction?

mikayla-maki commented:
> But, I have difficulty getting the image metadata to show up there.

You'll have to figure it out if you want to ship this PR :). I'd suggest looking at the cursor position UI, and learning how it's hooked into the rest of Zed. The key trait you'll want to understand is `StatusItemView`, and you can see how it's implemented for the CursorPosition struct here: https://github.com/zed-industries/zed/blob/2ff06cea6ee93b25e6d1d32ea247f70769207d82/crates/go_to_line/src/cursor_position.rs#L212-L232

I'm going to mark this as a draft for now, good luck and happy holidays!

kaf-lamed-beyt commented:
Thank you @mikayla-maki! I've been going through the CursorPosition implementation, actually. 

But, I'll go with your suggestion and approach everything carefully now. 

I'll make sure to mark this PR as ready when I'm set. Happy holidays!

kaf-lamed-beyt committed:
fix: use async Fs instead in favour of std::fs

The standard fs crate does filesystem operations synchronously which may lead to blocking the main thread when trying to access the image metadata, specifically fs.metadata to obtain the file size.

This PR addresses that issue, and also renders other image information like the dimensions (width & height), and the image type (PNG|JPG etc)

kaf-lamed-beyt committed:
Merge branch 'main' into worktree-abspath

kaf-lamed-beyt commented:
@mikayla-maki, kindly take a look when you get the chance. Thanks!

![Screenshot from 2024-12-29 15-36-54](https://github.com/user-attachments/assets/4a126d61-f3d9-460d-a5de-ebed450f62dd)

kaf-lamed-beyt commented:
Oh! I'll look into the spring issue and that divider in the UI too. 

I'd take a look at the settings crate to see how I can make the file meta unit customizable.

kaf-lamed-beyt commented:
People should now be able to set their preferred unit types. 

The image below uses the "decimal" unit
![image](https://github.com/user-attachments/assets/5150c1a3-2eb2-4b4a-b5bf-db3f93415e32). You can compare it with the previous screenshot I attached.

Here's how: 

![image](https://github.com/user-attachments/assets/07fc9670-4bed-4ef1-84e7-69403b45003f)

For the spacing issue @Angelk90, it is there by default. 

![image](https://github.com/user-attachments/assets/6be4372e-fae8-4bd0-812c-f4492065928c)

See an example of what it looks like in the current Zed version I have on my machine. The space between the `CursorPosition` and the language buffer is pretty much the same as what's in my first screenshot

kaf-lamed-beyt committed:
feat: allow users to chose their preferred units of measurements for image sizes

kaf-lamed-beyt committed:
Merge branch 'main' into worktree-abspath

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -52,9 +53,64 @@ pub struct ImageItem {
+            return Err(format!("File does not exist at path: {:?}", path));
+        }
+
+        let img = image::open(&path).map_err(|e| format!("Failed to open image: {}", e))?;
```
This should be wrapped in a `smol::unblock`, like here: https://github.com/zed-industries/zed/blob/2ff06cea6ee93b25e6d1d32ea247f70769207d82/crates/fs/src/fs.rs#L508-L509

kaf-lamed-beyt replied:
this too

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -52,9 +53,64 @@ pub struct ImageItem {
+        project: &Project,
+        cx: &AppContext,
+        fs: Arc<dyn fs::Fs>,
+    ) -> Result<(u32, u32, u64, &'static str), String> {
```
These should be a specific, named type rather than a tuple.

kaf-lamed-beyt replied:
addressed.

mikayla-maki commented on crates/image_viewer/src/image_info.rs:
```diff
@@ -0,0 +1,113 @@
+    width: Option<u32>,
+    height: Option<u32>,
+    file_size: Option<u64>,
+    color_type: Option<String>,
```
As mentioned below, these 4 values should have their own named struct. Then you can just have a single optional field for them all :)

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -143,9 +221,31 @@ impl ProjectItem for ImageItem {
-                    .await
+                    .await?;
+
+                let fs = Arc::new(fs::RealFs::default());
```
We don't want to construct a new `Fs` pointer here. There should be one in the project, that you can pass into `ImageStore::local` to do this. Or something similar.

kaf-lamed-beyt replied:
Yes. `project` has `fs`

Doing something like this now: 
```rs
let fs = project.read(cx).fs().clone();
```
would result in a type mismatch since `image_info` is expecting `cx` to be `AsyncAppContext` due to its async nature. 

But, `try_open` uses `cx: AppContext`. I'll try checking out `ImageStore::local` as you've suggested.

kaf-lamed-beyt replied:
I think this is the only part left now. @mikayla-maki, kindly take a look at the concern I raised here, thanks.

mikayla-maki replied:
You should be able to use `cx.update()` to get a version you can use to do `.read(cx)` :)

kaf-lamed-beyt replied:
oh! okay. I'll do this and let you know how it goes

kaf-lamed-beyt replied:
hi @mikayla-maki, suddenly this is preventing me from doing `cargo run`

```shell
error[E0277]: the trait bound `IndexMap<std::string::String, HighlightStyleContent>: JsonSchema` is not satisfied
   --> crates/theme/src/schema.rs:94:17
    |
94  |     pub syntax: IndexMap<String, HighlightStyleContent>,
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `JsonSchema` is not implemented for `IndexMap<std::string::String, HighlightStyleContent>`
    |
    = help: the following other types implement trait `JsonSchema`:
              &'a T
              &'a mut T
              ()
              (T0, T1)
              (T0, T1, T2)
              (T0, T1, T2, T3)
              (T0, T1, T2, T3, T4)
              (T0, T1, T2, T3, T4, T5)
            and 167 others
note: required by a bound in `insert_object_property`
   --> /home/.cargo/registry/src/index.crates.io-6f17d22bba15001f/schemars-0.8.21/src/_private.rs:126:43
    |
126 | pub fn insert_object_property<T: ?Sized + JsonSchema>(
    |                                           ^^^^^^^^^^ required by this bound in `insert_object_property`
    
```
    
In the theme crate
https://github.com/zed-industries/zed/blob/04518b11bc0c1e0959749ae66b318028944eeb60/crates/theme/src/schema.rs#L92-L95

This started happening when I pulled fixed the conflict I had before my recent commit.

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -143,9 +221,31 @@ impl ProjectItem for ImageItem {
+                if let Ok(()) = image_model.update(&mut cx, |image, cx| {
+                    let project_ref = project_clone.read(cx);
+
+                    if let Ok(metadata) = futures::executor::block_on(image.load_metadata(
```
This shouldn't be blocked on. Rather, return it out of the `.update()`, then `.await` it, then call `update()` again to assign the results.

kaf-lamed-beyt replied:
oh! thanks, i'll try this now

kaf-lamed-beyt replied:
@mikayla-maki, I did some cleanups. Kindly review and let me know what you think. 

The only issue I have is in this block. When I try to do it like you said &mdash; regarding this `executor::block_on` &mdash; I run into Rust lifetime issues several times.

mikayla-maki commented on crates/settings/src/settings.rs:
```diff
@@ -66,6 +66,8 @@ pub fn init(cx: &mut AppContext) {
         .unwrap();
     cx.set_global(settings);
+
+    ImageFileSizeUnitType::register(cx);
```
This shouldn't be registered here. This should be in an `init()` function in the `image_info` file.

kaf-lamed-beyt replied:
So for this scenario... i tried having an exact `init` function in `ImageItem` But I kept on running into an error. This one specifically. 

```shell
Thread "main" panicked with "unregistered setting type image_viewer::image_info::ImageFileSizeUnitType" at /oss/zed/crates/settings/src/settings_store.rs:341:32
   0: zed::reliability::init_panic_hook::{{closure}}
```

So to avoid it, I ended up using `call_once` from `std::sync` like this 👇🏼 so it doesn't call `::register()` everything the statusItem mounts

```rs
pub fn new(_workspace: &Workspace, cx: &mut AppContext) -> Self {
        static INIT: std::sync::Once = std::sync::Once::new();
        INIT.call_once(|| {
            ImageFileSizeUnitType::register(cx);
        });
 ```
 
 Lemme know if this is okay.

mikayla-maki commented on crates/settings/src/settings_store.rs:
```diff
@@ -238,6 +238,28 @@ trait AnySettingValue: 'static + Send + Sync {
 
+#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ImageFileSizeUnitType {
```
This shouldn't be here. It should be in the `image_info.rs` file.

mikayla-maki reviewed (changes requested):
Making a lot of progress! There's a couple things that need to be re-arranged though. You'll also need to get CI passing before we can merge this :)

That said, the feature itself is looking really good! Excited to have this.

kaf-lamed-beyt commented:
> Making a lot of progress! There's a couple things that need to be re-arranged though. You'll also need to get CI passing before we can merge this :)
> 
> That said, the feature itself is looking really good! Excited to have this.

Ayy! I'm happy to hear this @mikayla-maki! 

I'll go over the changes you've suggested and push tthem

kaf-lamed-beyt commented on crates/image_viewer/src/image_info.rs:
```diff
@@ -0,0 +1,113 @@
+    file_size: Option<u64>,
+    color_type: Option<String>,
+    _observe_active_image: Option<Subscription>,
+    image_unit_type: ImageFileSizeUnitType,
```
Awesome! Did this just now

kaf-lamed-beyt committed:
refactor

- create a standalaone type (ImageItemMeta) to hold image metadata
- add smol for async I/O actions when retrieving image properties
- move ImageFileSizeUnitType into image_info.rs
- used a call_once method from std::sync to register file size unit setting

kaf-lamed-beyt committed:
Merge branch 'main' into worktree-abspath

kaf-lamed-beyt committed:
Merge branch 'worktree-abspath' of github.com:kaf-lamed-beyt/zed into worktree-abspath

mikayla-maki commented:
Awesome! I'll mark this as draft for now. Let me know when it's next ready for review :)

kaf-lamed-beyt committed:
get fs from project instead of creating a new pointer

kaf-lamed-beyt committed:
update cargo lock

kaf-lamed-beyt committed:
Merge branch 'main' into worktree-abspath

smitbarmase committed:
fix block on

kaf-lamed-beyt committed:
Merge pull request #1 from 0xtimsb/fix-block-on

fix block on

kaf-lamed-beyt commented:
@mikayla-maki, can you take a look now?

jansol commented on crates/project/src/image_store.rs:
```diff
@@ -47,14 +48,107 @@ pub enum ImageStoreEvent {
+    match color_type {
+        ColorType::L8 => "Grayscale (8-bit)",
+        ColorType::La8 => "Grayscale with Alpha (8-bit)",
+        ColorType::Rgba8 => "PNG (32-bit color)",
+        ColorType::Rgb8 => "RGB (24-bit color)",
+        ColorType::Rgb16 => "RGB (48-bit color)",
+        ColorType::Rgba16 => "PNG (64-bit color)",
```
PNG is a *file* type that can contain many different color types. Knowing whether a file is PNG or JPEG or WEBP is important, but most image file types can contain many different color types so referring to the file type here is incorrect.

Looks like the ColorType of the image crate is not exactly ideal for this purpose. Some image formats explicitly differentiate e.g. `LUMINANCE_8` from `RED_8`, and `LUMINANCE_ALPHA_8` from `RG_8`. And then there are various different channel layouts, for example RGBA_4444 and RGB_565 for "16-bit color", or RGBA_1010102 which is 32 bits but people call it "10-bit color" because R,G,B each have 10 bits. `ExtendedColorType` has a bit more coverage so I'd use that if possible. The [PNG test suite](http://www.schaik.com/pngsuite/) contains files with various numbers of channels as well as bit depths ranging from 1 to 16 bits per channel as well as paletted images, I recommend using those to check the less common combinations.

I think the best option here is to simply return `"N channels, M bits per pixel"`. The status bar item should then show the *file* type and this channels+bpp info.

jansol reviewed (changes requested):
> I'm not too familiar with how GIFs work, but is it correct that we should see PNG (32-bit color) with a GIF open (maybe this refers to the format of each frame?)

No, GIF has its own encoding for the actual pixel data (and does not necessarily store full frames). Showing PNG is definitely wrong here. The file format needs to be pulled from the ImageReader with `.format()` *after* `.with_guessed_format()` (which sets the format) but *before* `.decode()` (which consumes the reader).

The channel count and bit depth are more complicated unfortunately.

kaf-lamed-beyt committed:
fix issue with generic color type mismatch in color_type_description

kaf-lamed-beyt commented:
Thanks @iamnbutler!

> I'd like to make this not a button, but I think it is easier for me to do that after this merges, as I should probably ship a variant of ButtonLike that matches the style of a button but is none-interactive (so the style stays consistent.) So no change here for now.

This is great! shipping a different variant of `Button`. I used it like so intending to add a tooltip &mdash; during the early stage of the PR instead of putting the info in a div. 

> I'm not too familiar with how GIFs work, but is it correct that we should see PNG (32-bit color) with a GIF open (maybe this refers to the format of each frame?)

And yes, you're right about this. Showing a format-specific description like `PNG` for a GIF for the color type `Rgba8` and `Rgba16` is misleading. GIFs can also fall within this color type, so inferring that it is PNG isn't good. My bad. 

Instead, we can leave it as `RGBA` for both. Thanks for the detailed breakdown, @jansol!

I pushed a commit addressing this now

iamnbutler commented:
Ok great! No blockers from me to merge this 👍

jansol commented:
> Instead, we can leave it as RGBA for both.

That's still problematic:
* a single-channel image can be "Grayscale" or "Luminance" or "Red" which are supposed to be treated differently
* a two-channel image could be "grayscale with alpha".... or it could be "red and green" (commonly used for non-color data e.g. in games)
* three channels could be RGB or YUV
* four channels could be RGBA, YUV with alpha or CMYK.
* there can be more channels than four (see "Deep Pixels" in OpenEXR)

(YUV is commonly used in video compression and occasionally still frames taken from videos may be kept in this format. JPEG files with CMYK color are used for professional printing. [The Windows Imaging Component documentation has a detailed summary.](https://learn.microsoft.com/en-us/windows/win32/wic/-wic-codec-native-pixel-formats))

ColorType is not a suitable data source since it silently expands 1, 2 and 4 bit formats to 8 bits. ExtendedColorType keeps the number of bits and adds an extra case for 8bpc CMYK and 8bpc alpha masks but still loses information about other types.  (it also can't handle >4 channel images but I don't know if image-rs can even read such images for now)

TL;DR: the only thing image-rs *somewhat* reliably tells us is the number of bits per pixel and the number of channels, so that's what I'd display in the status item, rather than claiming incorrect things about the meanings of the channels in an image.

kaf-lamed-beyt commented:
I see. 

So now, if I get you correctly, @jansol, you're suggesting we switch to `ExtendedColorType` instead of `ColorType`. am I right? 

Then, I'll have something like this: `GIF - 4 channels, 32 bits per pixel`? 

See the attached image. 

![image](https://github.com/user-attachments/assets/dfb02dee-550f-4153-8ac6-4f1d77e988ec)

My concern now is how do we make this descriptive enough ("grayscale" and whatnot)? I may be wrong, but I doubt many people would grasp this at first.

jansol commented:
> Then, I'll have something like this: GIF - 4 channels, 32 bits per pixel?

Precisely.

> My concern now is how do we make this descriptive enough ("grayscale" and whatnot)? I may be wrong, but I doubt many people would grasp this at first.

For now, not at all. If people aren't familiar with the concept of image channels making them look it up once is a much smaller problem in my book than making people who *do* understand it double-check every single image with some *other* tool because they can't trust what Zed tells them.

In the longer term I'd love to see image-rs to add a proper file metadata API so we can also know stuff like whether the image is encoded in sRGB or DCI-P3 or AdobeRGB or whatnot. But we should under no circumstances make up information that we don't have.

That said, `ExtendedColorType` *does* have special cases for CMYK (`Cmyk8`) and 8-bit alpha masks (`A8`) that could potentially be used to provide an extra bit of info for those cases with reasonable confidence that it'll be correct, but for others I wouldn't since that simply won't be reliable.

---

FWIW Here's what `file` and `ffprobe` respectively say about a random PNG I have lying around in my home directory:
```
file: PNG image data, 926 x 871, 8-bit/color RGB, non-interlaced
ffprobe: png, rgb24(pc, gbr/unknown/unknown), 926x871
```
And a JPEG right next to it:
```
file: JPEG image data, JFIF standard 1.01, resolution (DPI), density 72x72, segment length 16, progressive, precision 8, 415x424, components 3
ffprobe: mjpeg (Progressive), yuvj444p(pc, bt470bg/unknown/unknown), 415x424
```
As you can see, "RGB" is misleading here, even though both are in fact color images with 3 channels and image-rs does silently convert both to RGB after decoding them. But the point of the info view is to show information about the file on disk and not about what Zed is rendering to the screen.

kaf-lamed-beyt committed:
chore: consider image's appropriate bit depth/bit per pixel for the color type

instead of the former where i used ColorType in the pattern matching block to infer wrong values, this commit fixes that by using the ExtendedColorType for each image color type description considering the bit depth and channels.

it also includes a new property, format to annotate each image format, instead of relying on the file extensions, which may not be correct all the time.

kaf-lamed-beyt commented:
Yes, it is indeed misleading @jansol! thanks again for taking the time to elaborate on the concept behind the image channels/color type. 

I've also included the image format now.

kaf-lamed-beyt committed:
chore: update UnitFileSizeType enum to derive Binary as default instead of doing a manual impl

kaf-lamed-beyt committed:
cargo-machete don't complain about dependencies in use

kaf-lamed-beyt committed:
Merge branch 'main' into worktree-abspath

kaf-lamed-beyt committed:
remove duplicate imports

kaf-lamed-beyt commented:
CI passes now. 
cc: @mikayla-maki, @iamnbutler

iamnbutler commented:
hey @kaf-lamed-beyt, silly question but can you confirm after your latest changes the image information still shows up for you? I don't see it anymore.

![CleanShot 2025-01-23 at 09 59 15@2x](https://github.com/user-attachments/assets/a5f83913-9a0c-47f1-9730-94f31a0da256)

kaf-lamed-beyt commented:
Oh! Lemme do that now.

kaf-lamed-beyt commented:
so, on startup, when I opened it, there was an image currently open, the image info didn't show up. 

but, when I selected another one, it showed the image info. any idea what could be causing this @iamnbutler? Is there a way to ensure metadata loads on startup

![image](https://github.com/user-attachments/assets/1d59d677-d1d8-4344-b2f7-010056a0d8ae)

kaf-lamed-beyt committed:
fix: image format should not persist when there's no image view

kaf-lamed-beyt commented:
Yes. You're right. It is related to an image event not being called. Perhaps, I could try adding a new value `MetadataUpdated` or somn' in the `ImageItemEvent` enum. 

I'll try tinkering and let you know what I find.

kaf-lamed-beyt committed:
fix: load image metadata in image view when it renders for the first time

kaf-lamed-beyt commented:
@iamnbutler, I tried the `ImageItemEvent` approach, but it didn't yield positive results. 

After a while, I realized it was pretty much the same thing that goes into the `ProjectItem` implementation, and since the `new` method in `ImageView` is responsible for image loading on initial, render, I called, `ImageItem::image_info` in that block. 

It works. But, I don't know if it is optimal. Please take a look, thanks!

kaf-lamed-beyt committed:
Merge branch 'main' into worktree-abspath

kaf-lamed-beyt commented:
@mikayla-maki, I noticed that my PR had a merge conflict due to this update to GPUI (#22632)

I merged main into my branch (thinking about it now... I don't think I should've) and I've got tons of errors from crates that are unrelated to what I've worked on. Since `Model<T>` has been replaced with `Entity<T>`, I updated it in `image_info` but I can't find the replacement for `AsyncAppContext`. 

Care to walk me through? 

see an example error below: 

![image](https://github.com/user-attachments/assets/ec9fe1fe-be5c-4a49-b528-5562130a0fb0)

kaf-lamed-beyt committed:
chore: update implementation to use new GPUI3 APIs

kaf-lamed-beyt commented:
Hi @iamnbutler, kindly take a look at this. I think the CI checks are still waiting for approval to run.

kaf-lamed-beyt committed:
fix conflicts from main

kaf-lamed-beyt committed:
fix

mikayla-maki commented on crates/image_viewer/src/image_info.rs:
```diff
@@ -0,0 +1,147 @@
+    pub fn new(_workspace: &Workspace, cx: &mut App) -> Self {
+        static INIT: std::sync::Once = std::sync::Once::new();
+        INIT.call_once(|| {
+            ImageFileSizeUnitType::register(cx);
```
This should not be in a static. As mentioned previously, it should be added as a line in `image_viewer::init()`

kaf-lamed-beyt replied:
true. I guess, I left it that way because I couldn't find `init()`, I thought I'd have to write one myself. But it is already present in `image_viewer` as you mentioned.

mikayla-maki commented on crates/image_viewer/src/image_info.rs:
```diff
@@ -0,0 +1,147 @@
+}
+
+pub struct ImageInfo {
+    width: Option<u32>,
```
As mentioned before, this should use the `ImageItemMeta` struct with an `Option`, rather than defining each field individually. Like:

```rs
pub struct ImageInfo {
  metadata: Option<ImageItemMeta>,
  _observe_active_image: Option<Subscription>
}
```

kaf-lamed-beyt replied:
updated.

mikayla-maki commented on crates/image_viewer/src/image_viewer.rs:
```diff
@@ -31,10 +33,31 @@ impl ImageView {
+
+        if image_item.read(cx).image_meta.is_none() {
+            let image_item_clone = image_item.downgrade();
+            let project_clone = project.downgrade();
```
You can just call these `image_item` and `project`.

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -46,14 +49,143 @@ pub enum ImageStoreEvent {
 impl EventEmitter<ImageStoreEvent> for ImageStore {}
 
+#[derive(Clone, Debug)]
+pub struct ImageItemMeta {
```
Let's rename this to `ImageMetadata`

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -46,14 +49,143 @@ pub enum ImageStoreEvent {
+    ) -> Result<ImageItemMeta> {
+        let project_path = cx
+            .update(|cx| image.read(cx).project_path(cx))
+            .context("Failed to get project path")?;
```
The `update` call can only fail if the App has been closed, since this task was spawned.  This is a well established pattern in the Zed code base, so the extra error information here only clutters the code. 

Thank you for the diligence and properly handling these errors, but given the situation can remove this `context()` and just use `?` :)

kaf-lamed-beyt replied:
oh okay. thank you for clarifying! :)

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -46,14 +49,143 @@ pub enum ImageStoreEvent {
+                    .read(cx)
+                    .worktree_for_id(project_path.worktree_id, cx)
+            })
+            .context("Failed to get worktree")?
```
And this one.

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -46,14 +49,143 @@ pub enum ImageStoreEvent {
+
+        let worktree_root = cx
+            .update(|cx| worktree.read(cx).abs_path())
+            .context("Failed to get worktree root path")?;
```
And this one

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -46,14 +49,143 @@ pub enum ImageStoreEvent {
+
+        let image_path = cx
+            .update(|cx| image.read(cx).path().clone())
+            .context("Failed to get image path")?;
```
And this one

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -46,14 +49,143 @@ pub enum ImageStoreEvent {
+
+        let fs = project
+            .update(cx, |project, _| project.fs().clone())
+            .context("Failed to get filesystem")?;
```
And this one

mikayla-maki commented on crates/project/src/image_store.rs:
```diff
@@ -142,9 +274,17 @@ impl ProjectItem for ImageItem {
                     .update(&mut cx, |project, cx| project.open_image(path, cx))?
-                    .await
+                    .await?;
+                let image_metadata =
```
I'm noticing this image_metadata code is duplicated above in `ImageView::new`, could we add this as part of `Project::open_image`, so it always happens automatically?

kaf-lamed-beyt replied:
Huh. I think I tried this a while back, but it didn't yield positive results. I can't remember the exact function. I called `image_info` so we can always have the metadata as soon as it is available instead of on `active_pane_item` alone. 

I'll do this now, and let you know how it goes.

kaf-lamed-beyt replied:
Hi @mikayla-maki, so i'm facing a blocker here. This is how I'm trying to call `image_info` in `Project::open_image`

But, I'm running into this type mismatch where the context in the closure is `AsyncApp` in nature. But `image_item.read(cx)` expects `App`

```rs
let project = cx.entity();
        let task = self.image_store.update(cx, |image_store, cx| {
            image_store.open_image(path.into(), cx)
        });

        cx.spawn(|_, mut cx: AsyncAppContext| async move {
            let image_item = task.await?;
            if image_item.read(&cx).image_meta.is_none() {
                let metadata = ImageItem::image_info(image_item.clone(), project, &mut cx).await?;
                image_item.update(&mut cx, |image_item, _cx| {
                    image_item.image_meta = Some(metadata);
                })?;
            }
            Ok(image_item)
        })
 ```
 
What do you think I'm doing wrong?

mikayla-maki commented on crates/zed/src/zed.rs:
```diff
@@ -440,14 +443,12 @@ fn initialize_panels(
         };
 
         let (assistant_panel, assistant2_panel) = if is_assistant2_enabled {
-            log::info!("[assistant2-debug] initializing Assistant2");
```
I don't think we need to remove these logs.

kaf-lamed-beyt replied:
ah, this must have happened when I was trying to get `main` into my branch. I'll revert now.

mikayla-maki reviewed (changes requested):
Implementation looks great! I left a few notes about some code style changes and one bit of duplicated functionality I noticed. As soon as those are fixed I'd be happy to merge this :)

Thanks for your patience with this PR.

kaf-lamed-beyt commented:
> Implementation looks great! I left a few notes about some code style changes and one bit of duplicated functionality I noticed. As soon as those are fixed I'd be happy to merge this :)
> 
> Thanks for your patience with this PR.

Sure thing! I'll take a look at the requested changes now.

kaf-lamed-beyt committed:
refactor: code style

- rename ImageItemMeta to ImageMetadata
- update settings keymap to use "image_viewer" so we can pass additional settings la>
- move UnitType registration as ImageViewerSettings into image_viewer::init
- remove unnecessary context() for error handling

kaf-lamed-beyt committed:
load image metadata from open_open image

- update image_info to observe an image item not just the active one, so when there's image metadata available we update the image_meta property
- add a new value, MetadataUpdated to the ImageItemEvent enum to be used in the on_image_event function

kaf-lamed-beyt commented:
@mikayla-maki, everything works fine now. 

Kindly take a look. Thanks!

jansol commented:
@Angelk90 I don't think I'll have time to put together an exhaustive dataset any time soon.

@kaf-lamed-beyt thanks for seeing this through, although it ended up being a lot more work than the original issue suggested! (and sorry about that)

kaf-lamed-beyt committed:
remove unused dependencies

kaf-lamed-beyt commented:
CI is waiting for approval to run

cc: @mikayla-maki, @iamnbutler

kaf-lamed-beyt commented:
hi @iamnbutler, the tests keeps failing for MacOS and Linux. 

any idea how i can do a dry run of this CI on my machine before pushing?

iamnbutler commented:
Hey @kaf-lamed-beyt have you tried `script/clippy` from the zed dir?

iamnbutler commented:
but it looks like the tests are failing. Is `cargo test -p image_viewer` passing before you push?

kaf-lamed-beyt commented:
Huh. Lemme do that now.

kaf-lamed-beyt commented:
So i tried this now @iamnbutler and it was successful. No failures or somn'

kaf-lamed-beyt commented:
> but it looks like the tests are failing. Is `cargo test -p image_viewer` passing before you push? (or just cargo test)

Yeah. I'll do this too

kaf-lamed-beyt commented:
Ran into this error after doing `cargo test -p image_viewer` @iamnbutler 

```shell
error[E0425]: cannot find function `open_test_db` in crate `$crate`
   --> crates/image_viewer/src/image_viewer.rs:378:5
    |
378 | /     define_connection! {
379 | |         pub static ref IMAGE_VIEWER: ImageViewerDb<WorkspaceDb> =
380 | |             &[sql!(
381 | |                 CREATE TABLE image_viewers (
...   |
391 | |             )];
392 | |     }
    | |_____^ help: a function with a similar name exists: `open_main_db`
    |
   ::: /oss/zed/crates/db/src/db.rs:76:1
    |
76  |   async fn open_main_db<M: Migrator>(db_path: &Path) -> Option<ThreadSafeConnection<M>...
    |   ------------------------------------------------------------------------------------- similarly named function `open_main_db` defined here
    |
note: found an item that was configured out
   --> /oss/zed/crates/db/src/db.rs:99:14
    |
99  | pub async fn open_test_db<M: Migrator>(db_name: &str) -> ThreadSafeConn...
    |              ^^^^^^^^^^^^
note: the item is gated here
   --> /oss/zed/crates/db/src/db.rs:98:1
    |
98  | #[cfg(any(test, feature = "test-support"))]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the macro `define_connection` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0425`.
error: could not compile `image_viewer` (lib test) due to 1 previous error
```

mikayla-maki commented:
@kaf-lamed-beyt that error should be fine, let's not worry about it :)

Rather, the issue seems to be that you need to register your setting in some new places. In many of our tests, we start with a function that is similar to `init_test`, and what it does is call those `init()` methods we discussed earlier in order to register all of the things. You're going to need to go through the tests marked in CI, find those `init_test` methods, and then add a call to the `image_viewer::init` method :)

kaf-lamed-beyt commented:
Huh. OK, that makes sense now. I'll do this and let you know how it goes, @mikayla-maki!

Thanks for the heads up.

kaf-lamed-beyt committed:
add image_viewer init to tests

kaf-lamed-beyt commented:
Hi @mikayla-maki, i pushed a change just now. Kindly take a look.

kaf-lamed-beyt commented:
Out of 19 failures previously, now it is 1 failure. Good, good!

mikayla-maki commented:
@kaf-lamed-beyt You can run that test locally and see what the error is with cargo. `cargo test -p project test_image_not_loaded_twice` :)

kaf-lamed-beyt commented:
alright, i'll do this now

kaf-lamed-beyt commented:
Hi @mikayla-maki, I think it's still the same issue from before

```shell
cargo test -p image_viewer
   Compiling image_viewer v0.1.0 (/oss/zed/crates/image_viewer)
error[E0425]: cannot find function `open_test_db` in crate `$crate`
   --> crates/image_viewer/src/image_viewer.rs:378:5
    |
378 | /     define_connection! {
379 | |         pub static ref IMAGE_VIEWER: ImageViewerDb<WorkspaceDb> =
380 | |             &[sql!(
381 | |                 CREATE TABLE image_viewers (
...   |
391 | |             )];
392 | |     }
    | |_____^ help: a function with a similar name exists: `open_main_db`
    |
   ::: /oss/zed/crates/db/src/db.rs:76:1
    |
76  |   async fn open_main_db<M: Migrator>(db_path: &Path) -> Option<ThreadSafeConnection<M>...
    |   ------------------------------------------------------------------------------------- similarly named function `open_main_db` defined here
    |
note: found an item that was configured out
   --> /oss/zed/crates/db/src/db.rs:99:14
    |
99  | pub async fn open_test_db<M: Migrator>(db_name: &str) -> ThreadSaf...
    |              ^^^^^^^^^^^^
note: the item is gated here
   --> /oss/zed/crates/db/src/db.rs:98:1
    |
98  | #[cfg(any(test, feature = "test-support"))]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the macro `define_connection` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0425`.
error: could not compile `image_viewer` (lib test) due to 1 previous error
```

kaf-lamed-beyt committed:
add 'db/test-support' feature to image_viewer crate

kaf-lamed-beyt commented:
So @mikayla-maki, two things, i realized &mdash; after looking at the error closely &mdash; that the `open_test_db` is needs to work in a "test-support" scope, which is missing in the `image_viewer` crate, so i included it in `Cargo.toml` and ran the test again but this time around with the `--features test-support` &mdash; flag and it passed. 

I pushed a commit just now, so i the CI should pass now @iamnbutler

kaf-lamed-beyt commented:
@iamnbutler, same error from the CI 

```shell
──── STDERR:             project image_store::tests::test_image_not_loaded_twice
thread 'image_store::tests::test_image_not_loaded_twice' panicked at crates/project/src/image_store.rs:830:34:
called `Result::unwrap()` on an `Err` value: File does not exist at path: "/root/image_1.png"
```

I don't think I need to have `/root/image_1.png`, do I?

maxdeviant committed:
Merge branch 'main' into pr/21675

maxdeviant committed:
Sort dependencies

maxdeviant committed:
Don't remove unrelated lines

maxdeviant committed:
Merge branch 'main' into pr/21675

maxdeviant committed:
Do more cleanup

maxdeviant committed:
Inline constant that was only used once

maxdeviant committed:
More clean up

maxdeviant committed:
More clean up

maxdeviant committed:
Format

maxdeviant committed:
Do some more cleanup

maxdeviant committed:
Remove IO call

maxdeviant committed:
Remove more unnecessary I/O

maxdeviant committed:
Yet more cleanup

maxdeviant committed:
Consolidate updates

maxdeviant committed:
Build string representations at render time

maxdeviant committed:
Fix Clippy warnings

maxdeviant committed:
Format

maxdeviant committed:
Remove unneeded change from diff

maxdeviant committed:
Register settings first

maxdeviant committed:
Fix settings
